src/server/process_request.c

```diff
@@ -1654,3 +1654,37 @@ get_servername(unsigned int *port)
 
 	return name;
 }
+
+/**
+ * @brief
+ * 	Used to get serverer_instance_id which is of the form server_instance_name:server_instance_port
+ * 
+ * @return char *
+ * @return NULL - failure
+ * @retval !NULL - pointer to server_instance_id
+ */
+char *
+get_svr_inst_id()
+{
+	char svr_inst_name[PBS_MAXHOSTNAME + 1];
+	unsigned int svr_inst_port;
+	char *svr_inst_id = NULL;
+
+	/* server_instance_id is of the form "server_inst_name:svr_inst_port". 2 for  ":"" and NULL char */
+	svr_inst_id = malloc(PBS_MAXHOSTNAME + PBS_MAXPORTNUM + 2);
+	if (svr_inst_id == NULL)
+		return NULL;
+
+	if (gethostname(svr_inst_name, PBS_MAXHOSTNAME) == 0)
```

Comment:
the svr_inst_name should match its PSI value. I think when you use gethostname(), you will get the same svr_inst_name for all servers running on that host, even though in PSI the the server name values might be different. So, I think we should use PBS_SERVER_NAME value instead of the hostname here, which is what PSI uses.

Reply:
When it comes to the context of Multi-Server, PBS_SERVER refers to cluster name. This may or may not be a hostname of one of the server instances. It could totally be different from a hostname. 

Yes the idea is svr_inst_name should match with one of the PSI values. We also compare port numbers when all the server instances run on a same machine. So even if gethostname()/get_fullhostname() are the same in the above case, we also compare its port numbers, so it is not a problem. Also, Admin when configuring PBS_SERVER_INSTANCES he should make sure that he always either uses short hostname or FQDN to refer to a server instance.  

In case if we are using PBS_SERVER while populating pbs_conf.psi while only port number is provided by Admin for a server instance, we should change that to actually use the respective server's instance name. We are not supposed to use PBS_SERVER name here.

@subhasisb, please correct me If I am wrong.

Reply:
ok, can you please document this? That PSI should have valid hostnames, not PBS_SERVER (if it isn't documented already). Also, we should probably add a check inside pbs_loadconf() which verifies that each PSI value is a valid hostname.

Reply:
This also raises questions about how we'll handle talking to multiple servers in some of the IFL calls. Right now, we are in the process of adding multi-server support to IFLs like pbs_deljob, pbs_sigjob etc., which just take a jobid, which won't help identify the server if PSI will not be created using PBS_SERVER. So, if we do what you are proposing, then we'll need a different way to map jobids to servers.

Reply:
Actually, with this restriction, all servers will generate the same jobids, there will be job id conflicts, regardless of whether the servers run on the same host or on different hosts. I think your change cannot go in without fixing the jobid namespace conflict. Please let me know if I'm misunderstanding the implications of this.

Reply:
> ok, can you please document this? That PSI should have valid hostnames, not PBS_SERVER (if it isn't documented already). Also, we should probably add a check inside pbs_loadconf() which verifies that each PSI value is a valid hostname.

Yes we have not documented the above when PBS_SERVER_INSTANCES are used. Yes  I am planning to do exactly same thing i.e. add code changes to make sure PSI is a valid hostname. Also planning to document it. Don't know why we missed it earlier.

Reply:
> This also raises questions about how we'll handle talking to multiple servers in some of the IFL calls. Right now, we are in the process of adding multi-server support to IFLs like pbs_deljob, pbs_sigjob etc., which just take a jobid, which won't help identify the server if PSI will not be created using PBS_SERVER. So, if we do what you are proposing, then we'll need a different way to map jobids to servers.

Depending on jobid's hostname to figure out which is its server may or may not be the right idea. It is because going forward PBS_SERVER may or may not be hostname of one of the server instances configured. Instead, we can do the following.
An IFL API if it wants to find out which server is responsible for a given jobid, it tries to contact server instances in some random fashion and finds the right server. If the request reaches to a server which does not own the job it rejects with "unknown job id code". One of the server instances respond and process the request successfully. This way that IFL identifies which server it has to talk to. This is what @nithinj  is doing as part of his PR https://github.com/openpbs/openpbs/pull/2086

Reply:
> Actually, with this restriction, all servers will generate the same jobids, there will be job id conflicts, regardless of whether the servers run on the same host or on different hosts. I think your change cannot go in without fixing the jobid namespace conflict. Please let me know if I'm misunderstanding the implications of this.

The restriction to have same value for PBS_SERVER is not enforced in the code as of now. We need to keep this assumption in our minds and code it accordingly. So we are not using its(PBS_SERVER) value while configuring server instances and the code in this PR is exactly doing that. So given below is a mechanism how we can test as of now. 

If we make servers run on different complexes then we will get jobids differently because PBS_SERVER value is different. For example, we can get jobids like **1.spark,1.spark2, 2.spark, 2.spark2, 3.spark, 3.spark2**
In one of my test runs(as part of my PR testing) I have got the following
3.spark         root     workq    STDIN       50965   1   1    --    --  R 00:11
    --
3.spark2        root     workq    STDIN       73131   1   1    --    --  R 00:00
   Job run at Wed Nov 25 at 08:51 on (spark2:ncpus=1)
 
 
Also, when all server instances run on the same host, I think we can make aliases for PBS_SERVER in /etc/hosts and use it I believe,  in which case we get the jobids similar to the above.
 
Once we solve "jobid namespace problem" then we can strictly use only one value for PBS_SERVER across all complexes or across all server instances. In fact, we can enforce this in the code. When we do this we don't need to change any of the current code that is added as part of this PR.  This way I believe we don't need to wait till the completion of "jobid namespace feature".

Reply:
> An IFL API if it wants to find out which server is responsible for a given jobid, it tries to contact server instances in some random fashion and finds the right server. If the request reaches to a server which does not own the job it rejects with "unknown job id error code". One of the server instances(which is the owing server instance for the job) respond and process the request successfully. This way the IFL identifies which server it has to talk to. This is what @nithinj is doing as part of his PR #2086

Well that's not ideal for performance (both client and server side). Think about qdel list which was added recently .. we can't just broadcast the entire list of jobs to all servers, that will slow both client and server sides.

Reply:
The approach that I mentioned is being followed in the PR #2086 for most of the IFL APIs in general when they are called from a non scheduler like client/s. You can go through that PR for more details. 

When it comes to IFLs like pbs_deljob where performance matters we cannot follow the same approach. Instead, we can go through the following method to achieve the desired goal. There might be other better ways. Following are just my thoughts.
- Do a stat of the required jobs inside the IFL __pbs_deljoblist. Group the jobs by server_instance_id.
- Iterate over all the resultant server instances and pass one job-list at a time to the function PBSD_deljoblist().
- As the number of server instances configured are minimum this greatly enhances the performance.

Also, we are already following a slightly different approach for some IFLs like pbs_runjob for example when qrun with '-H' option is called. We actually do a stat for the given node through which we can find the right server instance id and send the request to that server instance directly instead of contacting servers in some random fashion and finding out which is the right server instance.

Reply:
Wait, I thought a runjob always went to the server currently owning the job.  The owning server is responsible to move and run the job.

Reply:
ya, we always send it to the job owner, I think Suresh is mistaken, but @nithinj can confirm once.

Reply:
sorry, I missed this thread earlier.
qrun sends the request to the job owning server. IFL will mention where the mother superior resides by doing a node_stat. It is a special case as we are dealing with two shared objects instead of one.